### PR TITLE
Rework pkg build into builders and provisioners

### PR DESCRIPTION
**Pull request by bauerm97**
**Description of the Pull Request (PR):**

Package build now consists of `Builders` and `Provisioners`. `Builder` takes different build request objects (json, deffile, io.Reader). `Provisioner` is created by a `builder` and is responsible for pulling from a source (dockerhub, shub, local image, tar archive, etc...) and converting it into a sandbox for stage two of the build process

Attn: @singularityware-admin

### CODE REVIEW

**Comment by bauerm97** [comment]
@tri-adam

File: pkg/image/sif.go [hunk 1]
```diff
@@ -12,25 +12,25 @@ import (
...
+func (i *SIF) Root() *specs.Root {
 	return &specs.Root{}
 }
 
-// isSif checks the "magic" of the given file and
-// determines if the file is of the sif type
-func isSif(path string) bool {
+// isSIF checks the "magic" of the given file and
+// determines if the file is of the SIF type
+func isSIF(path string) bool {
```
Comment (tri-adam) [inline pkg/image/sif.go:34]: How about an io.ReadSeeker as the input? Adds flexibility an allows us to minimize file handle open/closes.

File: pkg/image/sif.go [hunk 2]
```diff
@@ -12,25 +12,25 @@ import (
...
+// SIFFromSandbox converts the sandbox, s, to a SIF file
+func SIFFromSandbox(s *Sandbox) *SIF {
+	return &SIF{}
 }
 
-// SifFromPath returns a sif object of the file located at path
-func SifFromPath(path string) *sif {
-	return &sif{}
+// SIFFromPath returns a SIF object of the file located at path
+func SIFFromPath(path string) *SIF {
```
Comment (tri-adam) [inline pkg/image/sif.go:24]: How about an io.ReadSeeker as the input? Adds flexibility an allows us to minimize file handle open/closes.

File: pkg/build/builder.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"github.com/singularityware/singularity/pkg/image"
+)
+
+// Builder is
+type Builder struct {
+	Provisioner
+	Sandbox   image.Sandbox
+	Out       image.Image
```
Comment (tri-adam) [inline pkg/build/builder.go:18]: Any reason to not call this `Image`?

**Review by tri-adam — COMMENTED** [review]
Just a few comments...

**Pull request merged**